### PR TITLE
PCA SPMD python interfaces

### PR DESCRIPTION
**Pull request by jmhatre-work**
Python interfaces for PCA SPMD

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -471,7 +471,9 @@ def run(self):
         'onedal.datatypes',
         'onedal.common'
     ] + (['onedal.linear_model'] if ONEDAL_VERSION >= 20230100 else []) + (
-        ['onedal.spmd', 'onedal.spmd.linear_model'] if build_distribute else [])),
+        ['onedal.spmd',
+         'onedal.spmd.linear_model',
+         'onedal.spmd.decomposition'] if build_distribute else [])),
```
Comment (samir-nasibli) [inline setup.py:476]:
```suggestion
         'onedal.spmd.decomposition',
         'onedal.spmd.linear_model'] if build_distribute else [])),
```

File: examples/sklearnex/spmd_pca.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+#===============================================================================
+
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from numpy.testing import assert_allclose
+from onedal.spmd.decomposition import PCA as PCASpmd
+from sklearn.decomposition import PCA as PCAstock
+
+if __name__ == "__main__":
```
Comment (samir-nasibli) [inline examples/sklearnex/spmd_pca.py:25]: Please, simplify example, and avoid using `main`

File: onedal/__init__.py [hunk 3]
```diff
@@ -50,3 +50,4 @@
 
     if _is_dpc_backend:
         __all__.append('spmd.linear_model')
+        __all__.append('spmd.decomposition')
```
Comment (samir-nasibli) [inline onedal/__init__.py:53]:
```suggestion
        __all__ += ['spmd.decomposition', 'spmd.linear_model']
```

File: onedal/decomposition/pca.cpp [hunk 4]
```diff
@@ -136,10 +136,13 @@ ONEDAL_PY_INIT_MODULE(decomposition) {
 
     using task_list = types<task::dim_reduction>;
     auto sub = m.def_submodule("decomposition");
-
-    ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
-    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
-
+    #ifdef ONEDAL_DATA_PARALLEL_SPMD
+        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list_spmd, task_list);
+        ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
```
Comment (samir-nasibli) [inline onedal/decomposition/pca.cpp:141]: We don't support SPMD policy for infer?

Comment (ethanglaser) [reply onedal/decomposition/pca.cpp:141]: Just train - https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/algo/pca/detail/infer_ops_dpc.cpp

File: examples/sklearnex/spmd_pca.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===============================================================================
+
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from onedal.spmd.decomposition import PCA as PCASpmd
```
Comment (samir-nasibli) [inline examples/sklearnex/spmd_pca.py:20]: should be
```suggestion
from sklearnex.spmd.decomposition import PCA as PCASpmd
```

File: onedal/__init__.py [hunk 6]
```diff
@@ -49,4 +49,4 @@
     __all__ += ['basic_statistics', 'linear_model']
 
     if _is_dpc_backend:
-        __all__ += ['spmd.basic_statistics', 'spmd.linear_model']
+        __all__ += ["spmd.basic_statistics", "spmd.linear_model", "spmd.decomposition"]
```
Comment (samir-nasibli) [inline onedal/__init__.py:52]:
```suggestion
        __all__ += ['spmd.basic_statistics', 'spmd.decomposition', 'spmd.linear_model' ]
```

File: onedal/spmd/__init__.py [hunk 7]
```diff
@@ -14,4 +14,4 @@
 # limitations under the License.
 #===============================================================================
 
-__all__ = ['linear_model', 'basic_statistics']
+__all__ = ['linear_model', 'basic_statistics', 'decomposition']
```
Comment (samir-nasibli) [inline onedal/spmd/__init__.py:17]:
```suggestion
__all__ = ['basic_statistics', 'decomposition', 'linear_model']
```

File: setup.py [hunk 8]
```diff
@@ -476,8 +476,10 @@ def run(self):
          ) + (
         ['onedal.spmd',
          'onedal.spmd.linear_model',
-         'onedal.spmd.basic_statistics'
+         'onedal.spmd.basic_statistics',
+         'onedal.spmd.decomposition'
```
Comment (samir-nasibli) [inline setup.py:480]:
```suggestion
         'onedal.spmd.basic_statistics',
         'onedal.spmd.decomposition',
         'onedal.spmd.linear_model'
```

File: sklearnex/spmd/__init__.py [hunk 9]
```diff
@@ -14,4 +14,4 @@
 # limitations under the License.
 #===============================================================================
 
-__all__ = ['linear_model', 'basic_statistics']
+__all__ = ['linear_model', 'basic_statistics', 'decomposition']
```
Comment (samir-nasibli) [inline sklearnex/spmd/__init__.py:17]:
```suggestion
__all__ = ['basic_statistics', 'decomposition', 'linear_model']
```

File: onedal/decomposition/pca.cpp [hunk 10]
```diff
@@ -136,10 +136,13 @@ ONEDAL_PY_INIT_MODULE(decomposition) {
...
+    #else  
+        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
+        ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
+    #endif
```
Comment (samir-nasibli) [inline onedal/decomposition/pca.cpp:145]:
```suggestion
    #ifdef ONEDAL_DATA_PARALLEL_SPMD
        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list_spmd, task_list);
    #else
        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
    #endif // ONEDAL_DATA_PARALLEL_SPMD
    ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
```

File: examples/sklearnex/spmd_pca.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from sklearnex.spmd.decomposition import PCA as PCASpmd
+
+
+def get_data(data_seed, params_spmd):
+    drng = np.random.default_rng(data_seed)
+    X = drng.uniform(-7, 7, size=(params_spmd["ns"], params_spmd["nf"]))
+    y = drng.uniform(-7, 7, size=(params_spmd["ns"], params_spmd["nr"]))
```
Comment (KulikovNikita) [inline examples/sklearnex/spmd_pca.py:26]: PCA is a dimension reduction algorithm, thus it needs only `X` value.

File: examples/sklearnex/spmd_pca.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+# limitations under the License.
+# ===============================================================================
+
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from sklearnex.spmd.decomposition import PCA as PCASpmd
+
+
+def get_data(data_seed, params_spmd):
```
Comment (KulikovNikita) [inline examples/sklearnex/spmd_pca.py:23]: I believe you can hardcode parameters here

File: examples/sklearnex/spmd_pca.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+
+
+q = dpctl.SyclQueue("gpu")
+comm = MPI.COMM_WORLD
+rank = comm.Get_rank()
+size = comm.Get_size()
+
+
+params_spmd = {"ns": 15, "nf": 21, "nr": 23}
+X_spmd, y_spmd = get_data(0, params_spmd)
```
Comment (KulikovNikita) [inline examples/sklearnex/spmd_pca.py:37]: Data seed should depend on rank.

File: examples/sklearnex/spmd_pca.py (hunk 11 above)
Comment (jmhatre-work) [reply examples/sklearnex/spmd_pca.py:26]: https://github.com/intel/scikit-learn-intelex/blob/1674bb9719cc8cb42adce2dc8ef48ffe5f68b608/onedal/decomposition/pca.py#L46 

x, y both are kept for purpose of alignment. Should I make changes to remove y?

File: examples/sklearnex/spmd_pca.py [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+from mpi4py import MPI
+import dpctl
+from sklearnex.spmd.decomposition import PCA as PCASpmd
+
+
+def get_data(data_seed):
+    ns, nf, nr = 15, 21, 23
+    drng = np.random.default_rng(data_seed)
+    X = drng.uniform(-7, 7, size=(ns, nf))
+    y = drng.uniform(-7, 7, size=(ns, nr))
```
Comment (ethanglaser) [inline examples/sklearnex/spmd_pca.py:27]: I think you can remove all y data from this example as well

File: examples/sklearnex/spmd_pca.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from sklearnex.spmd.decomposition import PCA as PCASpmd
+
+
+def get_data(data_seed):
+    ns, nf = 15, 21
+    drng = np.random.default_rng(data_seed)
+    X = drng.uniform(-7, 7, size=(ns, nf))
```
Comment (Alexsandruss) [inline examples/sklearnex/spmd_pca.py:26]: Uniform distribution is theoretically meaningless for PCA: any components can be chosen.

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: examples/sklearnex/spmd_pca.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
+# ===============================================================================
+# Copyright 2023 Intel Corporation
```
Comment (samir-nasibli) [inline examples/sklearnex/spmd_pca.py:2]: Please rename your examples like `examples/skleaernex/<algo_name>_spmd.py`.

**Comment by samir-nasibli** [comment]
Hi just add GPU device and librariry requirements for your examples here https://github.com/intel/scikit-learn-intelex/blob/master/tests/run_examples.py#L127:L156:
```python
...
req_device['pca_spmd.py'] = ["gpu"]

...

req_library['pca_spmd.py''] = ['dpctl', 'mpi4py']
```
And also rename your examples and make `spmd` as suffix in the file names.

After that please run `/intelci: run`

File: onedal/__init__.py [hunk 17]
```diff
@@ -49,4 +49,4 @@
     __all__ += ['basic_statistics', 'linear_model']
 
     if _is_dpc_backend:
-        __all__ += ['spmd.basic_statistics', 'spmd.linear_model']
+        __all__ += ['spmd.basic_statistics', 'spmd.linear_model', 'spmd.decomposition']
```
Comment (samir-nasibli) [inline onedal/__init__.py:52]:
```suggestion
        __all__ += ['spmd.basic_statistics', 'spmd.decomposition', 'spmd.linear_model',]
```

File: setup.py [hunk 18]
```diff
@@ -475,9 +475,11 @@ def run(self):
           ] if ONEDAL_VERSION >= 20230100 else []
          ) + (
         ['onedal.spmd',
-         'onedal.spmd.linear_model',
-         'onedal.spmd.basic_statistics'
+         'onedal.spmd.basic_statistics',
+         'onedal.spmd.decomposition',
+         'onedal.spmd.linear_model'
          ] if build_distribute else [])),
+
```
Comment (samir-nasibli) [inline setup.py:482]:
```suggestion
```

**Review by samir-nasibli — APPROVED** [review]
LGTM

**Review by Alexsandruss — APPROVED** [review]

File: examples/sklearnex/pca_spmd.py [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+import numpy as np
+from mpi4py import MPI
+import dpctl
+from sklearnex.spmd.decomposition import PCA
+
+
+def get_data(data_seed):
+    ns, nf = 15, 21
+    drng = np.random.default_rng(data_seed)
+    X = drng.random(size=(ns, nf))
```
Comment (Alexsandruss) [inline examples/sklearnex/pca_spmd.py:26]: Random is mostly the same as uniform - doesn't make sense for PCA.
@KulikovNikita, do you think this number of samples is enough for example?

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

**Review by Alexsandruss — APPROVED** [review]
